### ISSUE TITLE
Review request - tensors not involved in operations will not be involved in backprop?

### ISSUE THREAD

**Issue by EricLBuehler** [huggingface/candle#1232]
Hello everybody,

I just wanted to confirm explicitly that if a `Tensor` (not a `Var`) is involved in an operation, it will not be involved in the backprop, and so is "frozen"? Thank you!

**Comment by LaurentMazare** [comment]
I'm not sure to understand exactly what involved/frozen mean. A `Tensor` can actually be a variable - in which case it behaves like a `Var` (actually it's more that `Var` is a wrapper around such `Tensor`s), but this only happens for tensors that are initialized as variables via some very specific functions. Tensors created through normal operations will be involved in the backprop: a gradient will be calculated and will flow back to the `Var` that have been used to create this tensor but its value will not be updated by the optimizers so in a sense it will be frozen.

**Comment by EricLBuehler** [comment]
Ok. How can I ensure that no gradients are calculated for the Tensor? I want the weights to be untrainable (frozen).

**Comment by LaurentMazare** [comment]
Similar to PyTorch, you have a `.detach` method that will cut the backprop graph where it is applied (so no gradient will flow back to this part).

**Comment by LaurentMazare** [comment]
Yes, that's the goal though obviously the semantics can be a bit tricky if a var is used multiple times some of them are detached some not, but if it's all detached then the var should not be updated during the optimizer step. If you notice something weird about it, certainly report it (ideally with a small repro).

**Comment by EricLBuehler** [comment]
Ok, thanks.

As an example, is the following implementation valid (found [here](https://github.com/EricLBuehler/candle-lora/blob/d4ee8de886ba316cb3082630216f4d58d6c3acd8/candle-lora/src/frozenlinear.rs#L25-L34))?

```rust
Self::new(
    old.weight().detach()?,
    match old.bias() {
        Some(bias) => Some(bias.detach()?),
        None => None,
    },
)
```

Just as a final confirmation, even if the weights are `.detach`ed, because the storage is shared, the original weights will still track gradients, and it is important to not use the original weights after "freezing" the layer. Here is the implementation of `.detach`, for future reference:

> https://github.com/huggingface/candle/blob/6c990a33ea4635bf98b180f6e4c99e6795ccfbab/candle-core/src/tensor.rs#L1810-L1821
